File: Java version/app/src/main/java/it/unical/mat/parsers/asp/dlv2/DLV2ParserBaseVisitorImplementation.java
package it.unical.mat.parsers.asp.dlv2;

import it.unical.mat.parsers.asp.ASPDataCollection;
import java.util.HashMap;
import org.antlr.v4.runtime.BailErrorStrategy;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.ConsoleErrorListener;
import org.antlr.v4.runtime.DefaultErrorStrategy;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.atn.PredictionMode;

public class DLV2ParserBaseVisitorImplementation extends DLV2ParserBaseVisitor <Void> {
	private final ASPDataCollection answerSets;
	private HashMap <Integer, Integer> costs;
    
    private DLV2ParserBaseVisitorImplementation(final ASPDataCollection answerSets) {
        this.answerSets = answerSets;
    }

    @Override
    public Void visitAnswer_set(DLV2Parser.Answer_setContext ctx) { 
        answerSets.addAnswerSet();
        
        if(ctx.cost() != null && !ctx.cost().isEmpty()) {
        	costs = new HashMap <> ();
        	final String[] firstCost = ctx.cost().COST_LABEL().getText().split(" ")[1].split("@");
        	
        	costs.put(Integer.parseInt(firstCost[1]), Integer.parseInt(firstCost[0]));
        } 
        
        if(costs != null)
        	costs.forEach((level, cost) -> answerSets.storeCost(level, cost));
        
        return visitChildren(ctx);
    }

    @Override
    public Void visitLevel(DLV2Parser.LevelContext ctx) {
    	final int level = Integer.parseInt(ctx.INTEGER(1).getText()), cost = Integer.parseInt(ctx.INTEGER(0).getText());
    	
    	costs.put(level, cost);
    	answerSets.storeCost(level, cost);
    	
        return null;
    }

    @Override
    public Void visitPredicate_atom(DLV2Parser.Predicate_atomContext ctx) {
        answerSets.storeAtom(ctx.getText());

        return null;
    }

    public static void parse(final ASPDataCollection answerSets, final String atomsList, final boolean two_stageParsing) {
    	final CommonTokenStream tokens = new CommonTokenStream(new DLV2Lexer(CharStreams.fromString(atomsList)));
        final DLV2Parser parser = new DLV2Parser(tokens);
        final DLV2ParserBaseVisitorImplementation visitor = new DLV2ParserBaseVisitorImplementation(answerSets);
        
        if(!two_stageParsing) {
        	visitor.visit(parser.output());
        	
        	return;
        }
        
        parser.getInterpreter().setPredictionMode(PredictionMode.SLL); 
        parser.removeErrorListeners();
        parser.setErrorHandler(new BailErrorStrategy());
     
        try {
        	visitor.visit(parser.output());
        } catch (final RuntimeException exception) {
        	if(exception.getClass() == RuntimeException.class && exception.getCause() instanceof RecognitionException) {
        		tokens.seek(0);
        		parser.addErrorListener(ConsoleErrorListener.INSTANCE);
        		parser.setErrorHandler(new DefaultErrorStrategy());
        		parser.getInterpreter().setPredictionMode(PredictionMode.LL); 
                visitor.visit(parser.output());
        	}
        }
    }
}

File: CSharp version/it/unical/mat/parsers/asp/dlv2/DLV2ParserBaseVisitorImplementation.cs
using System;
using System.Collections.Generic;
using Antlr4.Runtime;
using Antlr4.Runtime.Atn;

namespace it.unical.mat.parsers.asp.dlv2
{
    public class DLV2ParserBaseVisitorImplementation : DLV2ParserBaseVisitor<object>
    {
        private readonly IASPDataCollection answerSets;
        private Dictionary<int, int> costs;

        private DLV2ParserBaseVisitorImplementation(IASPDataCollection answerSets)
        {
            this.answerSets = answerSets;
        }

        public override object VisitAnswer_set(DLV2Parser.Answer_setContext context)
        {
            answerSets.AddAnswerSet();

            if (context.cost() != null && !context.cost().IsEmpty)
            {
                costs = new Dictionary<int, int>();
                string[] FirstCost = context.cost().COST_LABEL().GetText().Split(' ')[1].Split('@');

                costs[Int32.Parse(FirstCost[1])] = Int32.Parse(FirstCost[0]);
            }

            if (costs != null)
                foreach (KeyValuePair<int, int> entry in costs)
                    answerSets.StoreCost(entry.Key, entry.Value);

            return VisitChildren(context);
        }

        public override object VisitLevel(DLV2Parser.LevelContext context)
        {
            int level = Int32.Parse(context.INTEGER()[1].GetText()), cost = Int32.Parse(context.INTEGER()[0].GetText());

            costs[level] = cost;
            answerSets.StoreCost(level, cost);

            return null;
        }

        public override object VisitPredicate_atom(DLV2Parser.Predicate_atomContext context)
        {
            answerSets.StoreAtom(context.GetText());

            return null;
        }

        public static void Parse(IASPDataCollection answerSets, string atomsList, bool two_stageParsing)
        {
            CommonTokenStream tokens = new CommonTokenStream(new DLV2Lexer(CharStreams.fromstring(atomsList)));
            DLV2Parser parser = new DLV2Parser(tokens);
            DLV2ParserBaseVisitorImplementation visitor = new DLV2ParserBaseVisitorImplementation(answerSets);

            if (!two_stageParsing)
            {
                visitor.Visit(parser.output());

                return;
            }

            parser.Interpreter.PredictionMode = PredictionMode.SLL;

            parser.RemoveErrorListeners();

            parser.ErrorHandler = new BailErrorStrategy();

            try
            {
                visitor.Visit(parser.output());
            }
            catch (SystemException exception)
            {
                if (exception.GetBaseException() is RecognitionException)
                {
                    tokens.Seek(0);
                    parser.AddErrorListener(ConsoleErrorListener<object>.Instance);

                    parser.ErrorHandler = new DefaultErrorStrategy();
                    parser.Interpreter.PredictionMode = PredictionMode.LL;

                    visitor.Visit(parser.output());
                }
            }
        }
    }
}